pkg/controller/clustersync/clustersync_controller.go

```diff
@@ -0,0 +1,844 @@
+			return reconcile.Result{}, err
+		}
+	}
+
+	if needToDoFullReapply {
+		logger.Info("setting last full apply time")
+		lease.Spec.RenewTime = metav1.NowMicro()
+		if needToCreateLease {
+			logger.Info("creating lease for ClusterSync")
+			lease.Namespace = leaseKey.Namespace
+			lease.Name = leaseKey.Name
+			lease.OwnerReferences = []metav1.OwnerReference{{
+				APIVersion:         hivev1.SchemeGroupVersion.String(),
+				Kind:               "ClusterSync",
+				Name:               clusterSync.Name,
+				UID:                clusterSync.UID,
+				BlockOwnerDeletion: pointer.BoolPtr(true),
+			}}
+			if err := r.Create(context.Background(), lease); err != nil {
+				logger.WithError(err).Log(controllerutils.LogLevel(err), "could not create lease for ClusterSync")
+				return reconcile.Result{}, err
+			}
+		} else {
+			logger.Info("updating lease for ClusterSync")
+			if err := r.Update(context.Background(), lease); err != nil {
+				logger.WithError(err).Log(controllerutils.LogLevel(err), "could not update lease for ClusterSync")
+				return reconcile.Result{}, err
+			}
+		}
+	}
+
+	result := reconcile.Result{Requeue: true, RequeueAfter: r.timeUntilFullReapply(lease)}
+	if syncSetsNeedRequeue || selectorSyncSetsNeedRequeue {
+		result.RequeueAfter = 0
+	}
+	return result, nil
+}
+
+func (r *ReconcileClusterSync) applySyncSets(
+	syncSetType string,
+	syncSets []syncset.CommonSyncSet,
+	syncStatuses []hiveintv1alpha1.SyncStatus,
+	needToDoFullReapply bool,
+	resourceHelper resource.Helper,
+	logger log.FieldLogger,
+) (newSyncStatuses []hiveintv1alpha1.SyncStatus, requeue bool) {
+	// Sort the syncsets to a consistent ordering. This prevents thrashing in the ClusterSync status due to the order
+	// of the syncset status changing from one reconcile to the next.
+	sort.Slice(syncSets, func(i, j int) bool {
+		return syncSets[i].AsMetaObject().GetName() < syncSets[j].AsMetaObject().GetName()
+	})
+
+	for _, syncSet := range syncSets {
+		logger := logger.WithField(syncSetType, syncSet.AsMetaObject().GetName())
+		oldSyncStatus, indexOfOldStatus := getOldSyncStatus(syncSet, syncStatuses)
+		// Remove the matching old sync status from the slice of sync statuses so that the slice only contains sync
+		// statuses that have not been matched to a syncset.
+		if indexOfOldStatus >= 0 {
+			last := len(syncStatuses) - 1
+			syncStatuses[indexOfOldStatus] = syncStatuses[last]
+			syncStatuses = syncStatuses[:last]
+		}
+
+		// Determine if the syncset needs to be applied
+		switch {
+		case needToDoFullReapply:
+			logger.Debug("applying syncset because it is time to do a full re-apply")
+		case indexOfOldStatus < 0:
+			logger.Debug("applying syncset because the syncset is new")
+		case oldSyncStatus.Result != hiveintv1alpha1.SuccessSyncSetResult:
+			logger.Debug("applying syncset because the last attempt to apply failed")
+		case oldSyncStatus.ObservedGeneration != syncSet.GetStatus().ObservedGeneration:
```

Comment:
note to self: so far, still no need to have `status.observedGeneration` on [s]ss. The meta generation would do fine here.

Reply:
observedGeneration ensures that the resource identifiers in `status.resources` is up to date.